### ISSUE TITLE
Interactive process buffers and terminal emulation

### ISSUE THREAD

**Issue by refi64** [howl-editor/howl#218]
Idea: currently, if you run a process that requires input, you'll usually end up with some sort of EOF error. Maybe Howl could allow interactive processes somehow?

**Comment by shalabhc** [comment]
:+1: I've wanted this too.

Actually I've wanted a full terminal emulator but that's may be too much too ask...

**Comment by kirbyfan64** [comment]
@shalabhc https://developer.gnome.org/vte/unstable/VteTerminal.html

Seems to be a terminal emulator widget of sorts. Unfortunately, it seems to be 99% undocumented...

**Comment by shalabhc** [comment]
Interesting. I'm not sure if this is exactly what I want or is this a black box?

Basically I want a buffer to be a terminal, where I can run a shell and other programs that read input. As a more advanced feature I would like the terminals to be shared across Howl instances, so the terminal state has to be decoupled from the Howl process itself. This makes it somewhat like tmux, I suppose.

(Even sharing non-interactive external processes across Howl instances is useful, IMO).

**Comment by etosan** [comment]
Recently I started using `howl` more again and rembered how great UI it has. Aullar is pretty good monospace characters drawing engine, it seems. I remember already having discussion with @nilnor about possiblity of "lifting" Aullar "out" as standalone renderer, but conclusion was it's too tied to `howl` and it would require lot of work to decouple (that I am unable to do).

Anyway, I believe Aullar already has everything you would need to implement minimal terminal emulator.  Rendering is one of the harder parts of terminal emulator. I also think it would be cool if there could be Aullar terminal emulator :).

I strongly would be against any VTE "swallowing". I would rather have no emulator than have VTE. And even if you don't want to hear my opinion I will say it: I believe VTE is especially bad, maybe even sh*ttiest emulator on the planet, and should not be used for anything else besides toys and "macsy" `Terminal` ripoffs. It's also pretty big chunk of GTK code as well, and I am not sure how exactly would that work inside of `howl`. As far as I understand, `howl` draws it's window itself completely, so how would it hand off drawing some part of it's window rectangle to VTE widget? Probably doable, but then renderer would be VTE's one, completely decoupled from `howl`'s own character drawing subsystem. This would be Scintilla "situation" once again. I don't think VTE would be proper way to go about it, given `howl`'s "spirit" (completely custom renderer).

I grepped `howl` codebase for traditional pseudotty and child process management and found nothing.  It seems some higher order GTK-ish API is used to create sub-processes for external "commands".

These APIs take care of creating the pipes to read and write from and into child process, and also seem to take care of the forking. Those are fine for collecting output of child into the buffer, but not really sufficient to emulate terminal. Real "terminal" has more to it.

What would be needed to create barely minimal terminal? 

These three things:
  - pty/tty layer - this is called pseudo teletype and emulates actual console to a running child program (this is how any terminal, xterm, VTE or anything does it's magic) and is "provided" by OS/kernel. 
  - child signal management is used to "communicate" with child. Most important would be sending SIGWINCH when `howl`'s "terminal" view changes size. There are few others singals that need to be sent, but this signal sending set could be kept pretty minimal.
  - actual ANSI escape sequences parser, heart of the emulator. More on this later.

First point: Pseudoteletype seems daunting at first, but if you want to play, `luapty` package used to do the trick. `howl` would create pty and would get slave device which it then needs to hand out to the child it forks. This is what makes child programs print in "colors". Color printing detection works roughly like this: child asks isatty(stdout), if it returns true, it's a real terminal and child can use colors, else child reverts to plain text. Naturally, in current setup, isatty(stdout) returns false, because stdout of child is really a pipe created by subprocess API. Very simple "forwarder" that just creates pty spawns a child and forwards everything child spits to it's own parent terminal is not that many lines of C or lua. It's just simple read()/write() pump. That's because we are "abusing" forwarder's parent terminal to actually "draw" the output of the child. More on that in the last part.

Second point: Signal management. This is not that bad either. First, children cannot, or rather will not, really, just send signals to our `howl` "parent". At minimum, only thing `howl` would need to do, is to inform teletype device of it's size and size throwing changes (aka Aullar dimensions in cells/characters ). This is done with ioctl() calls on pty master that `howl` would be holding ownership of anyway. This way window resizes can be handled. Implementing only window resizes however, would rob us of ability to terminate misbehaving child right from `howl`. By using same icotl() interface, support for sending CTRL+C (or any `howl` bind really), eg. support for sending SIGINT (aka CTRL+C) can be implemented. Implementing these two signals only, would already get you quite usable "terminal". There are other signals terminals can send to children, but that could be added later and are not that important. One thing, that would need to be done here is handling of graceful exit of child (eg when it terminates). By default OS would inform `howl` itself about child termination by sending it SIGCHLD. Normally this signal is ignored by `howl` and would do nothing. Howl would need to be extended to handle it. With this, you can now have "long running" children, that will think that they are running in real terminal (that has colors and size). Think shell, vi, nano, or emacs even :).

Third point: ANSI escape parser. This that "more on that later part". This is meat and potatoes of the terminal magic. When I said that pty can be handled by just dumb read()/write() pump I meant it. Story now is, what are you gonna do with bytes you read out the child. Currently, those are just simply pasted into the buffer, as plain text. What would be needed though, is to analyze every incoming byte and decide whether it means a character (plain text) or it is ANSI drawing command. Plain text is just drawn on the spot, then advancing the cursor. But commands would need to be decoded. Then, depending on command, cursor has to be moved around the Aullar rectangle, or text colors have to change ("of the writing brush"). This is the hardest part, which makes or breaks the terminal emulator. 

So far we only talked about reading from the child, which is more complex task, but converse is true as well. To be able to use the child interactively, writing to it is required. Just writing bytes into the child will make it think it's reading plain text from keyboard. Input "commands" however, need to be ANSI encoded as well. For example, to pass the information to the child that user pressed a right arrow key to advance the cursor right, this has to be appropriately encoded into proper ANSI sequence and only this sequence can be written to a child. Child reading this sequence, will then think that user actually pressed the right arrow.  It will process it and send move cursor right ANSI command back to us. This way it all ties together.

Good thing is that more advanced terminals tend to implement superset of commands supported by simpler ones. So theoretically simplest one bearable could be found and implemented initially, then extended.

What are prerequisites that would be needed to be fixed make this worth the time of all the involved:
  - split window modes (vertical/horizontal panels at best arbitrarily nested) 
    - terminal view/buffer would need to be decoupled from normal buffers to be usable anyway
    - it needs to be focusable as well
  - decoupled multi-window (currently it is not possible to have multiple windows with own view(s))

In multiwindow case, decoupling views from "main" window, would allow `howl` to have separately movable windows. With help of window manager one can have standalone terminal (or edit buffer) next to other. With splits, workflow allows user to have one part of the split UI dedicated to file and other to terminal (in project directory for example) to run commands in. Implementing both enforces proper separation of view from being hardcoded to singular main window. 

Once this is completed, `howl` is ready have pty+signals and rudimentary ANSI emulation throwing into the mix.

Typing it like this shows its quite a lot of work but not entirely impossible. But yeah it seems a bit unreasonable. With careful planning, I think it could be done. I have a hunch moon script could be leveraged effectively to simplify lot of this, but I still suck at it.

This is how terminal widget is done in, for example, IDEA Java IDE. It is, relatively, minimal emulator capable of parsing and sending ANSI sequences written completely in Java.

That is also all, that you would need, to have colors in terminal proper :).  

There are some halfway through hacks, like forcing color output in some tools (like ls for example) with explicit command line switches, but then you still have the problem of parsing and decoding ANSI color sequences tool will emit in such mode. If you are wiling to do that ANSI parsing, it's much worth to parse it from proper pty, than just hack it on top of sub-process pipe.

Final words are: I still think any builtin `howl` native terminal would be way better than VTE. It would be also much more integrated and configurable. Backlog could be fetched from other buffers with `howl` commands, text could be sent into child program from buffers as well, and terminal handling could be exteneded with moonscript (and lua).

**Comment by nilnor** [comment]
Sorry for the very long silence here.

Howl development has been quiet on `master` for years, while work continued on the `next` branch — that branch has now been merged. Master has changed substantially since the 0.6 release: a full port to GTK 4, and a rewrite of the command line and interaction APIs, among a lot else.

We're closing out the old backlog, and feature requests are being closed for good rather than parked. Howl is developed at a pace that doesn't support an open wishlist, and leaving these sitting here indefinitely implies a roadmap that doesn't exist. That's not a comment on the merit of the idea — thank you for taking the time to write it up.

Please don't reopen this one. Bug reports against current master are still welcome.
